test/src/main/java/org/apache/accumulo/test/functional/ManagerAssignmentIT.java

```diff
@@ -88,9 +119,283 @@ public void test() throws Exception {
+      TabletLocationState ondemand = getTabletLocationState(c, tableId);
+      assertNull(ondemand.future);
+      assertNull(ondemand.current);
+      assertEquals(flushed.current, ondemand.last);
+
+      // put it back online
+      c.tableOperations().online(tableName, true);
+      online = getTabletLocationState(c, tableId);
+      assertNull(online.future);
+      assertNotNull(online.current);
+      assertEquals(online.current, online.last);
+
+    }
+  }
+
+  @Test
+  public void testScannerAssignsOneOnDemandTablets() throws Exception {
+    try (AccumuloClient c = Accumulo.newClient().from(getClientProps()).build()) {
+      String tableName = super.getUniqueNames(1)[0];
+      c.tableOperations().create(tableName);
+      String tableId = c.tableOperations().tableIdMap().get(tableName);
+      loadDataForScan(c, tableName);
+
+      TreeSet<Text> splits = new TreeSet<>();
+      splits.add(new Text("f"));
+      splits.add(new Text("m"));
+      splits.add(new Text("t"));
+      c.tableOperations().addSplits(tableName, splits);
+      c.tableOperations().onDemand(tableName, true);
+      assertTrue(c.tableOperations().isOnDemand(tableName));
+
+      // Wait 2x the TabletGroupWatcher interval for ondemand
+      // tablets to be unassigned.
+      Thread.sleep(10000);
+
+      List<TabletStats> stats = getTabletStats(c, tableId);
+      // There should be no tablets online
+      assertEquals(0, stats.size());
+      assertEquals(0, TabletLocator.getLocator((ClientContext) c, TableId.of(tableId))
+          .onDemandTabletsOnlined());
+
+      c.tableOperations().clearLocatorCache(tableName);
+
+      Range scanRange = new Range("a", "c");
+      Scanner s = c.createScanner(tableName);
+      s.setRange(scanRange);
+      // Should return keys for a, b, c
+      assertEquals(3, Iterables.size(s));
+
+      stats = getTabletStats(c, tableId);
+      // There should be one tablet online
+      assertEquals(1, stats.size());
+      assertEquals(1, TabletLocator.getLocator((ClientContext) c, TableId.of(tableId))
+          .onDemandTabletsOnlined());
+
+    }
+  }
+
+  @Test
+  public void testScannerAssignsMultipleOnDemandTablets() throws Exception {
+    try (AccumuloClient c = Accumulo.newClient().from(getClientProps()).build()) {
+      String tableName = super.getUniqueNames(1)[0];
+      c.tableOperations().create(tableName);
+      String tableId = c.tableOperations().tableIdMap().get(tableName);
+      loadDataForScan(c, tableName);
+
+      TreeSet<Text> splits = new TreeSet<>();
+      splits.add(new Text("f"));
+      splits.add(new Text("m"));
+      splits.add(new Text("t"));
+      c.tableOperations().addSplits(tableName, splits);
+      c.tableOperations().onDemand(tableName, true);
+      assertTrue(c.tableOperations().isOnDemand(tableName));
+
+      // Wait 2x the TabletGroupWatcher interval for ondemand
+      // tablets to be unassigned.
+      Thread.sleep(10000);
```

Comment:
Could do a while loop here if possible too.

Reply:
Addressed in 8738077